lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,538 @@
+    /// for a new element to be inserted.
+    fn push(&mut self, value: I) {
+        assert!(self.has_space(&value));
+        self.allocated_bytes += value.size_of();
+        self.elements.push(value);
+    }
+}
+
+#[pin_project]
+pub struct Batcher<St, Prt, T>
+where
+    Prt: Partitioner,
+{
+    /// The timer to maintain periodic flushes from this `Batcher`
+    timer: T,
+    /// The maximum number of items that are allowed per-batch
+    batch_item_limit: usize,
+    /// The total number of bytes a single batch in this struct is allowed to
+    /// hold. If None effectively unlimited.
+    batch_allocation_limit: usize,
+    /// The store of live batches. Note that the key here is an option type,
+    /// on account of the interface of `Prt`.
+    batches: HashMap<Option<Prt::Key>, Batch<Prt::Item>, BuildHasherDefault<XxHash64>>,
+    /// The store of 'closed' batches. When this is not empty it will be
+    /// preferentially flushed prior to consuming any new items from the
+    /// underlying stream.
+    closed_batches: Vec<(Option<Prt::Key>, Vec<Prt::Item>)>,
+    /// The partitioner for this `Batcher`
+    partitioner: Prt,
+    #[pin]
+    /// The stream this `Batcher` wraps
+    stream: St,
+}
+
+impl<St, Prt, T> Batcher<St, Prt, T>
+where
+    St: Stream + Unpin,
+    Prt: Partitioner + Unpin,
+{
+    pub fn new(
+        stream: St,
+        partitioner: Prt,
+        timer: T,
+        batch_item_limit: NonZeroUsize,
+        batch_allocation_limit: Option<NonZeroUsize>,
+    ) -> Self {
+        Self {
+            batch_item_limit: batch_item_limit.get(),
+            batch_allocation_limit: batch_allocation_limit
+                .map_or(usize::max_value(), NonZeroUsize::get),
+            batches: HashMap::default(),
+            closed_batches: Vec::default(),
+            timer,
+            partitioner,
+            stream,
+        }
+    }
+}
+
+impl<St, Prt, T> Stream for Batcher<St, Prt, T>
+where
+    St: Stream + Unpin,
+    Prt: Partitioner + Unpin,
+    Prt::Item: ByteSizeOf + From<St::Item>,
+    T: Timer,
+{
+    type Item = (Option<Prt::Key>, Vec<Prt::Item>);
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.stream.size_hint()
+    }
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
+            if !this.closed_batches.is_empty() {
+                return Poll::Ready(this.closed_batches.pop());
+            }
+            match this.stream.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    if this.timer.has_elapsed() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        this.timer.reset();
```

Comment:
This batch timeout scheme is too simplistic, I believe.

Right now, you're treating batch timeouts as a global event.  However, the code this is going to replace deals with timeouts at the batch level.  Maybe there's something to be said for timing out globally, but it will definitely result in batches that do not wait their full `batch.timeout_secs` before sending.

Reply:
cc @jszwedko @lukesteensen @bruceg if any of y'all have opinions on the above (flushing all batches on an even cadence vs flushing them individually)

Reply:
I'm open to either approach. The challenge that I see with the per-batch timeout is we'll periodically need to scan all batches to determine which have gone past their timer and need to be serviced, plus the minimal overhead of a timer per batch etc. It struck me as desirable to reduce the amount of extra state needed considering how hot this path is likely to be in practice, even if it does mean some batches go out smaller than they might otherwise have done. 

No partition's batch will go out _later_ than its timeout though, which seems the more vital constraint to satisfy.

Reply:
Chatting about this further externally, I think we've agreed that the change in behavior is small enough to warrant the vastly simpler implementation that we've achieved here.